logical/framework/field_data.go

```diff
@@ -291,6 +293,41 @@ func (d *FieldData) getPrimitive(k string, schema *FieldSchema) (interface{}, bo
 		}
 		return result, true, nil
 
+	case TypeHeader:
+		// First try to parse this as a map
+		var mapResult map[string][]string
+		if err := mapstructure.WeakDecode(raw, &mapResult); err == nil {
+			// The http.Header Get method is case-insensitive in the key it takes,
+			// but searches the map for a matching canonicalized value. Most headers
+			// will arrive in this format, but just in case they don't, let's
+			// canonicalize all the keys.
+			result := http.Header{}
+			for k, slice := range mapResult {
+				canonicalKey := textproto.CanonicalMIMEHeaderKey(k)
```

Comment:
Do you need to do this step?  It seems to me you would want to preserve formatting and any comparisons you would retrieve the header using the `Get` function and do any normalized text compares there.

Reply:
It's definitely weird.

The reason I do it is because if you send in a header of `"hello:world"` and then later do `header.Get("hello")`, it _doesn't_ find `"world"`. That's because under the hood, the `Get` method first converts `"hello"` to `"Hello"` and _then_ looks for it in the map. 

I thought this was a really weird behavior, and tried to think through how to best achieve what someone would _expect_. I thought about, instead of using an `http.Header`, maybe I should just use a straight-out map. But then, people are sending _in_ headers so maybe they would expect that kind of behavior.

In the end, I did this because I thought from a _usability_ perspective it might be the best option. But I'm open to other approaches because it's a tough call.

Reply:
I don't know if I explained that well LOL. In the example of `"hello:world"` I was thinking of '"hello"' as the key and `"world"` as the value. Probably would have been a better example if I'd done JSON or something.

I was trying to say that if you give it a lower-case key of hello, then later try to pull out a lower-case hello, it finds nothing. And it's because under the hood, it converts it to Title Case before searching for it in the map, so it's looking for `"Hello"` while it's sitting in the map as `"hello"`, and thus misses it. It converts it to Title Case using that canonical method.

So by also calling that canonical method on keys on the way in, it also Title Cases the keys in the map so they'll have hits. This is definitely a trade-off because, while you'll now have case-insensitive hits using the `header.Get("hello")` method, if you need to pull headers out like _this_: header["hello"]`` it won't work anymore. You'd use the `Get` method for headers that you know only have one value, and the map-key-accessing method for headers that may have multiple values.

Ultimately, I'm honestly not super in love with the magic the `http.Header` type is doing because I think it's actually pretty weird. This is why I'm tempted to just turn it into a `map[string][]string`, and to do away with the canonical stuff, so it's way more evident and straight-forward how to use it.

Reply:
I too struggled with this type for whatever reason, it is challenging to use.  Maybe you're doing a little too much sanitation and you just need to pass through what they give you and serialize it directly to disk when necessary.  Here is the code I came up with dealing with headers, config, and output. https://github.com/hashicorp/vault/blob/master/vault/ui.go

Reply:
Probably you shouldn't do any sanitization at all -- you should get it to the correct type and just copy it over verbatim. There's no need for you to do any textproto stuff when Header is doing it anyways.